plugin/sampling/strategystore/adaptive/processor.go

```diff
@@ -0,0 +1,506 @@
+// Copyright (c) 2018 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adaptive
+
+import (
+	"errors"
+	"io"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/cmd/collector/app/sampling/model"
+	ss "github.com/jaegertracing/jaeger/cmd/collector/app/sampling/strategystore"
+	jio "github.com/jaegertracing/jaeger/pkg/io"
+	"github.com/jaegertracing/jaeger/plugin/sampling/internal/calculationstrategy"
+	"github.com/jaegertracing/jaeger/plugin/sampling/internal/leaderelection"
+	"github.com/jaegertracing/jaeger/storage/samplingstore"
+	"github.com/jaegertracing/jaeger/thrift-gen/sampling"
+)
+
+const (
+	maxSamplingProbability = 1.0
+
+	getThroughputErrMsg = "failed to get throughput from storage"
+
+	defaultFollowerProbabilityInterval = 20 * time.Second
+
+	// The number of past entries for samplingCache the leader keeps in memory
+	serviceCacheSize = 25
```

Comment:
yes we only use the most recent two entries in the serviceCache. We can probably remove the older entries in the cache. I'll add a TODO for now because there might've been a reason my younger and smarter self did this.